File: julia/test/semantic_index_compilation_foundation_test.jl
const SEMANTIC_INTROSPECTION_FIXTURE_ROOT = normpath(
    joinpath(@__DIR__, "..", "..", "capability_conformance", "semantic_introspection"),
)

function _semantic_fixture_source(name)
    return read(joinpath(SEMANTIC_INTROSPECTION_FIXTURE_ROOT, "$(name).spec"), String)
end

function _captured_semantic_outcome_error(call)
    try
        call()
    catch error
        @test error isa SemanticIndexError
        return error
    end
    @test false
    return nothing
end

function _semantic_outcome_plain_json(value)
    if value === nothing || value isa Bool || value isa Integer || value isa AbstractString
        return true
    elseif value isa AbstractVector
        return all(_semantic_outcome_plain_json, value)
    elseif value isa AbstractDict
        return all(key isa AbstractString for key in keys(value)) &&
               all(_semantic_outcome_plain_json, values(value))
    end
    return false
end

function _semantic_outcome_keys(value)
    found = Set{String}()
    if value isa AbstractDict
        for (key, item) in value
            push!(found, String(key))
            union!(found, _semantic_outcome_keys(item))
        end
    elseif value isa AbstractVector
        for item in value
            union!(found, _semantic_outcome_keys(item))
        end
    end
    return found
end

@testset "Semantic index compiled-or-failed foundation" begin
    graph_source = _semantic_fixture_source("graph")
    graph_index = semantic_index(
        graph_source;
        logical_name = "graph.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )

    expected_snapshot = SemanticSnapshot(
        "snapshot:0",
        SemanticCompiledSnapshotState,
        false,
        SemanticSourceTextDetail,
        true,
    )
    @test semantic_snapshot(graph_index) == expected_snapshot
    @test Base.hash(expected_snapshot, UInt(0)) ==
          Base.hash(semantic_snapshot(graph_index), UInt(0))
    @test to_json(semantic_snapshot(graph_index)) == Dict{String,Any}(
        "id" => "snapshot:0",
        "state" => "compiled",
        "has_execution" => false,
        "source_detail_ceiling" => "text",
        "content_digest_available" => true,
    )
    @test repr(graph_index) ==
          "SemanticIndex(source_id=\"source:0\", snapshot_state=\"compiled\", source_detail_ceiling=\"text\", has_execution=false)"
    @test !occursin("graph.spec", repr(graph_index))
    @test !occursin("Top", repr(graph_index))

    expected_authority = SemanticCompilationAuthority(true, true, true)
    @test compilation_authority(graph_index) == expected_authority
    @test Base.hash(expected_authority, UInt(0)) ==
          Base.hash(compilation_authority(graph_index), UInt(0))
    @test to_json(compilation_authority(graph_index)) == Dict{String,Any}(
        "parsed" => true,
        "validated" => true,
        "compiled" => true,
    )
    @test compilation_diagnostic(graph_index) === nothing

    expected_entry = SemanticEntrySelection("Top", "first_authored_marker")
    @test entry_selection(graph_index) == expected_entry
    @test Base.hash(expected_entry, UInt(0)) ==
          Base.hash(entry_selection(graph_index), UInt(0))
    @test to_json(entry_selection(graph_index)) == Dict{String,Any}(
        "label" => "Top",
        "basis" => "first_authored_marker",
    )

    expected_plan = SemanticGeneratedPlanInput(
        contract_id = "linkedspec-generated-source-v2",
        format_version = 2,
        source_identity = "graph.spec",
        rows = (
            SemanticGeneratedPlanRow("Top", "and_acode_seq"),
            SemanticGeneratedPlanRow("Child", "rep_acode"),
        ),
    )
    @test generated_plan_input(graph_index) == expected_plan
    @test Base.hash(expected_plan, UInt(0)) ==
          Base.hash(generated_plan_input(graph_index), UInt(0))
    @test to_json(generated_plan_input(graph_index)) == Dict{String,Any}(
        "contract_id" => "linkedspec-generated-source-v2",
        "format_version" => 2,
        "source_identity" => "graph.spec",
        "rows" => Any[
            Dict{String,Any}("label" => "Top", "family" => "and_acode_seq"),
            Dict{String,Any}("label" => "Child", "family" => "rep_acode"),
        ],
    )
    @test expected_plan.rows isa Tuple
    @test_throws MethodError push!(expected_plan.rows, SemanticGeneratedPlanRow("Other", "default"))

    plan_json = to_json(generated_plan_input(graph_index))
    plan_json["source_identity"] = "mutated.spec"
    plan_json["rows"][1]["label"] = "Mutated"
    @test generated_plan_input(graph_index) == expected_plan

    none_index = semantic_index(
        graph_source;
        logical_name = "graph-none.spec",
        source_detail_ceiling = SemanticSourceNoneDetail,
    )
    @test semantic_snapshot(none_index) == SemanticSnapshot(
        "snapshot:0",
        SemanticCompiledSnapshotState,
        false,
        SemanticSourceNoneDetail,
        false,
    )
    plan_ceiling_error = _captured_semantic_outcome_error(
        () -> generated_plan_input(none_index),
    )
    @test plan_ceiling_error.stage == "apply_source_ceiling"
    @test plan_ceiling_error.code == "semantic_source_detail_forbidden"

    explicit_index = semantic_index(
        graph_source;
        logical_name = "graph-explicit.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
        entry_rule = "Child",
    )
    @test entry_selection(explicit_index) ==
          SemanticEntrySelection("Child", "explicit_selector")
    @test generated_plan_input(explicit_index).source_identity == "graph-explicit.spec"

    markerless_index = semantic_index(
        "First:\n /x/\n\nSecond:\n /y/\n";
        logical_name = "markerless.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
    )
    @test entry_selection(markerless_index) ==
          SemanticEntrySelection("First", "first_authored_rule")

    calls_index = semantic_index(
        _semantic_fixture_source("calls_and_staging");
        logical_name = "calls_and_staging.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
    )
    @test LinkedSpecJulia._semantic_authored_definition_order(calls_index) == (
        (kind = "function", name = "normalize", line = 1),
        (kind = "rule", name = "Top", line = 3),
        (kind = "rule", name = "Done", line = 9),
    )
    @test [row.label for row in generated_plan_input(calls_index).rows] == ["Top", "Done"]
    @test [row.family for row in generated_plan_input(calls_index).rows] == ["default", "default"]

    failed_index = semantic_index(
        _semantic_fixture_source("failed");
        logical_name = "failed.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )
    @test semantic_snapshot(failed_index).state == SemanticFailedCompilationSnapshotState
    @test compilation_authority(failed_index) == SemanticCompilationAuthority(true, false, false)
    expected_failure = SemanticCompilationDiagnostic(
        code = "bare_edge_target_undefined",
        stage = "normalize_edges",
        message = "bare edge in rule 'Top' targets undefined rule 'Missing'",
        fields = ("rule_label" => "Top", "target" => "Missing"),
    )
    @test compilation_diagnostic(failed_index) == expected_failure
    @test Base.hash(expected_failure, UInt(0)) ==
          Base.hash(compilation_diagnostic(failed_index), UInt(0))
    @test entry_selection(failed_index) === nothing
    @test generated_plan_input(failed_index) === nothing

    failure_json = to_json(compilation_diagnostic(failed_index))
    @test failure_json["fields"] == Dict{String,Any}(
        "rule_label" => "Top",
        "target" => "Missing",
    )
    failure_json["fields"]["target"] = "Mutated"
    @test Dict(compilation_diagnostic(failed_index).fields)["target"] == "Missing"

    parse_failure = semantic_index(
        "Top:::\n /x/\n";
        logical_name = "parse-failure.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
    )
    @test compilation_authority(parse_failure) == SemanticCompilationAuthority(false, false, false)
    @test compilation_diagnostic(parse_failure).code == "semantic_index_parse_failed"
    @test compilation_diagnostic(parse_failure).stage == "parse_source"
    @test Dict(compilation_diagnostic(parse_failure).fields) == Dict{String,Any}("line" => 1)

    empty_failure = semantic_index(
        "";
        logical_name = "empty.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
    )
    @test compilation_authority(empty_failure) == SemanticCompilationAuthority(true, false, false)
    @test compilation_diagnostic(empty_failure) == SemanticCompilationDiagnostic(
        code = "no_rules_defined",
        stage = "validate_spec",
        message = "spec does not define any rules",
    )

    missing_entry = semantic_index(
        "Top::\n /x/\n";
        logical_name = "missing-entry.spec",
        source_detail_ceiling = SemanticSourceIdentityDetail,
        entry_rule = "Missing",
    )
    @test compilation_authority(missing_entry) == SemanticCompilationAuthority(true, true, false)
    @test compilation_diagnostic(missing_entry) == SemanticCompilationDiagnostic(
        code = "entry_rule_not_found",
        stage = "select_entry_rule",
        message = "entry rule 'Missing' is not defined",
        fields = ("entry_rule" => "Missing",),
    )
    @test entry_selection(missing_entry) === nothing
    @test generated_plan_input(missing_entry) === nothing

    compile_fallback = LinkedSpecJulia._semantic_language_diagnostic(
        CompiledSpecException("compile fallback"),
        "semantic_index_compilation_failed",
        "compile_source",
    )
    @test compile_fallback.code == "semantic_index_compilation_failed"
    @test compile_fallback.stage == "compile_source"
    @test compile_fallback.message == "compile fallback"
    plan_fallback = LinkedSpecJulia._semantic_language_diagnostic(
        ArgumentError("plan fallback"),
        "semantic_index_generated_plan_failed",
        "build_generated_plan",
    )
    @test plan_fallback.code == "semantic_index_generated_plan_failed"
    @test plan_fallback.stage == "build_generated_plan"
    @test plan_fallback.message == "ArgumentError: plan fallback"
    @test LinkedSpecJulia._semantic_is_fatal_exception(InterruptException())
    @test LinkedSpecJulia._semantic_is_fatal_exception(OutOfMemoryError())
    @test LinkedSpecJulia._semantic_is_fatal_exception(StackOverflowError())
    @test !LinkedSpecJulia._semantic_is_fatal_exception(ErrorException("ordinary"))

    bytes_index = semantic_index(
        Vector{UInt8}(codeunits(graph_source));
        logical_name = "graph.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )
    @test semantic_snapshot(bytes_index) == semantic_snapshot(graph_index)
    @test compilation_authority(bytes_index) == compilation_authority(graph_index)
    @test entry_selection(bytes_index) == entry_selection(graph_index)
    @test generated_plan_input(bytes_index) == generated_plan_input(graph_index)

    runtime_index = semantic_index(
        _semantic_fixture_source("runtime");
        logical_name = "runtime.spec",
        source_detail_ceiling = SemanticSourceTextDetail,
    )
    @test semantic_snapshot(runtime_index).state == SemanticCompiledSnapshotState
    @test semantic_snapshot(runtime_index).has_execution === false
    @test generated_plan_input(runtime_index).rows ==
          (SemanticGeneratedPlanRow("Top", "rep_acode"),)

    public_json = Any[
        to_json(semantic_snapshot(graph_index)),
        to_json(compilation_authority(graph_index)),
        to_json(entry_selection(graph_index)),
        to_json(generated_plan_input(graph_index)),
        to_json(compilation_diagnostic(failed_index)),
    ]
    @test all(_semantic_outcome_plain_json, public_json)
    forbidden_keys = Set([
        "ast",
        "body_ast",
        "compiled_spec",
        "definition_order",
        "descriptor",
        "path",
        "rules_by_label",
        "source_bytes",
        "source_text",
    ])
    @test isempty(reduce(union, (_semantic_outcome_keys(value) for value in public_json)) ∩ forbidden_keys)
    @test isdefined(LinkedSpecJulia, :semantic_query)
    @test !isdefined(LinkedSpecJulia, :semantic_records)

    implementation = read(
        joinpath(@__DIR__, "..", "src", "semantic", "SemanticCompilationOutcome.jl"),
        String,
    )
    @test length(collect(eachmatch(r"parse_spec_with_staged_user_function_definitions\(source_text\)", implementation))) == 1
    @test length(collect(eachmatch(r"validate_spec\(parsed\)", implementation))) == 1
    @test length(collect(eachmatch(r"compile_spec\(parsed; validate_source = false\)", implementation))) == 1
    @test length(collect(eachmatch(r"resolve_entry_rule\(candidate, options\.entry_rule\)", implementation))) == 1
    @test length(collect(eachmatch(r"build_generated_rule_plan\(candidate\)", implementation))) == 1
    for forbidden in (
        "SpecLoader",
        "load_spec(",
        "runtime_parse(",
        "runtime_execute(",
        "execute_generated_parser",
        "LinkedSpecTraceEmitter",
        "diagnostic_output_sink",
        "semantic_observation",
        "to_descriptor_json",
    )
        @test !occursin(forbidden, implementation)
    end
    @test occursin("error isa InterruptException", implementation)
    @test occursin("error isa OutOfMemoryError", implementation)
    @test occursin("error isa StackOverflowError", implementation)
end

File: lua/test/semantic_index_compilation_foundation_test.lua
-- FUTURE-PARITY-BACKLOG.10.7.2.2 — staged compiled-or-failed foundation.

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label or "assertion failed" end
end

local function check_equal(actual, expected, label)
  check(actual == expected, (label or "values differ") ..
    ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
end

local function check_contains(actual, expected, label)
  check(tostring(actual):find(expected, 1, true) ~= nil, (label or "text differs") ..
    ": expected to contain " .. expected .. ", got " .. tostring(actual))
end

local function read_file(path)
  local handle = assert(io.open(path, "rb"))
  local value = assert(handle:read("*a"))
  assert(handle:close())
  return value
end

local linkedspec = require("linkedspec")
local json = linkedspec.json
local source_emitter = linkedspec.source_emitter

local function options(logical_name, source_detail_ceiling, entry_rule)
  return {
    logical_name = logical_name,
    source_detail_ceiling = source_detail_ceiling,
    entry_rule = entry_rule,
  }
end

local function check_same_json(actual, expected, label)
  check_equal(json.encode(actual), json.encode(expected), label)
end

local function capture_error(operation, label)
  local ok, value = pcall(operation)
  check_equal(ok, false, label .. " rejected")
  return ok and nil or value
end

local function check_opaque(value, kind, label)
  check_equal(tostring(value), kind, label .. " stable kind")
  check_equal(getmetatable(value), "protected", label .. " protected metatable")
  check_equal(next(value), nil, label .. " empty raw state")
  local count = 0
  for _ in pairs(value) do count = count + 1 end
  check_equal(count, 0, label .. " empty iteration")
  check_equal(pcall(function() value.leak = "/tmp/private.spec" end), false,
    label .. " write rejected")
  check_equal(pcall(function() setmetatable(value, {}) end), false,
    label .. " metatable replacement rejected")
end

local function plain_json(value, active)
  local value_type = type(value)
  if value == nil or value_type == "boolean" or value_type == "number" or value_type == "string" then
    return true
  end
  if value_type ~= "table" then return false end
  active = active or {}
  if active[value] then return false end
  active[value] = true
  for key, item in pairs(value) do
    if type(key) ~= "string" and type(key) ~= "number" then return false end
    if not plain_json(item, active) then return false end
  end
  active[value] = nil
  return true
end

local function collect_keys(value, found, active)
  found = found or {}
  active = active or {}
  if type(value) ~= "table" or active[value] then return found end
  active[value] = true
  for key, item in pairs(value) do
    if type(key) == "string" then found[key] = true end
    collect_keys(item, found, active)
  end
  active[value] = nil
  return found
end

local graph = read_file("capability_conformance/semantic_introspection/graph.spec")
local graph_index = linkedspec.semantic_index(graph, options("graph.spec", "text"))

local snapshot = graph_index:semantic_snapshot()
check_same_json(snapshot:to_json(), json.harray({
  id = "snapshot:0",
  state = "compiled",
  has_execution = false,
  source_detail_ceiling = "text",
  content_digest_available = true,
}), "graph snapshot")
check_equal(snapshot.id, "snapshot:0", "snapshot id")
check_equal(snapshot.state, "compiled", "snapshot state")
check_equal(snapshot.has_execution, false, "snapshot execution absence")

local authority = graph_index:compilation_authority()
check_same_json(authority:to_json(), json.harray({
  parsed = true,
  validated = true,
  compiled = true,
}), "graph compilation authority")
check_equal(graph_index:compilation_diagnostic(), nil, "graph diagnostic absence")

local entry = graph_index:entry_selection()
check_same_json(entry:to_json(), json.harray({
  label = "Top",
  basis = "first_authored_marker",
}), "graph entry selection")

local plan = graph_index:generated_plan_input()
check_same_json(plan:to_json(), json.harray({
  contract_id = "linkedspec-generated-source-v2",
  format_version = 2,
  source_identity = "graph.spec",
  rows = json.array({
    json.harray({ label = "Top", family = "and_acode_seq" }),
    json.harray({ label = "Child", family = "rep_acode" }),
  }),
}), "graph generated plan")

local detached_rows = plan.rows
detached_rows[1].label = "Mutated"
detached_rows[2] = nil
check_equal(graph_index:generated_plan_input().rows[1].label, "Top", "plan rows detached")
check_equal(#graph_index:generated_plan_input().rows, 2, "plan row count retained")
local detached_plan_json = plan:to_json()
detached_plan_json.source_identity = "/private/mutated.spec"
detached_plan_json.rows[1].family = "default"
check_equal(graph_index:generated_plan_input().source_identity, "graph.spec", "plan JSON identity detached")
check_equal(graph_index:generated_plan_input().rows[1].family, "and_acode_seq", "plan JSON rows detached")

check_equal(
  tostring(graph_index),
  "SemanticIndex(source_id=\"source:0\", snapshot_state=\"compiled\", " ..
    "source_detail_ceiling=\"text\", has_execution=false)",
  "compiled index string"
)
for _, forbidden in ipairs({ "graph.spec", "Top", "/tmp", "table:", "userdata:" }) do
  check_equal(tostring(graph_index):find(forbidden, 1, true), nil,
    "compiled index string redacts " .. forbidden)
end

local none = linkedspec.semantic_index(graph, options("graph-none.spec", "none"))
check_equal(none:semantic_snapshot().content_digest_available, false, "none snapshot digest absence")
local plan_ceiling = capture_error(function() return none:generated_plan_input() end, "none plan ceiling")
check_equal(plan_ceiling.stage, "apply_source_ceiling", "none plan ceiling stage")
check_equal(plan_ceiling.code, "semantic_source_detail_forbidden", "none plan ceiling code")

local explicit = linkedspec.semantic_index(graph, options("graph-explicit.spec", "identity", "Child"))
check_same_json(explicit:entry_selection():to_json(), json.harray({
  label = "Child",
  basis = "explicit_selector",
}), "explicit entry selection")
check_equal(explicit:generated_plan_input().source_identity, "graph-explicit.spec",
  "explicit plan identity")

local markerless = linkedspec.semantic_index(
  "First:\n /x/\n\nSecond:\n /y/\n",
  options("markerless.spec", "identity")
)
check_equal(markerless:entry_selection().label, "First", "markerless entry label")
check_equal(markerless:entry_selection().basis, "first_authored_rule", "markerless entry basis")

local calls = linkedspec.semantic_index(
  read_file("capability_conformance/semantic_introspection/calls_and_staging.spec"),
  options("calls_and_staging.spec", "identity")
)
check_equal(calls:semantic_snapshot().state, "compiled", "staged calls compile")
check_equal(calls:semantic_snapshot().has_execution, false, "staged calls do not execute")
check_equal(calls:generated_plan_input().rows[1].label, "Top", "staged calls first row")
check_equal(calls:generated_plan_input().rows[1].family, "default", "staged calls first family")
check_equal(calls:generated_plan_input().rows[2].label, "Done", "staged calls second row")
check_equal(calls:generated_plan_input().rows[2].family, "default", "staged calls second family")

local failed = linkedspec.semantic_index(
  read_file("capability_conformance/semantic_introspection/failed.spec"),
  options("failed.spec", "text")
)
check_equal(failed:semantic_snapshot().state, "failed_compilation", "native failure snapshot")
check_same_json(failed:compilation_authority():to_json(), json.harray({
  parsed = true,
  validated = false,
  compiled = false,
}), "native failure authority")
local diagnostic = failed:compilation_diagnostic()
check_same_json(diagnostic:to_json(), json.harray({
  code = "bare_edge_target_undefined",
  stage = "normalize_edges",
  message = "bare edge in rule 'Top' targets undefined rule 'Missing'",
  fields = json.harray({ rule_label = "Top", target = "Missing" }),
}), "native validation diagnostic")
check_equal(failed:entry_selection(), nil, "native failure entry absence")
check_equal(failed:generated_plan_input(), nil, "native failure plan absence")
local detached_fields = diagnostic.fields
detached_fields.target = "Mutated"
check_equal(failed:compilation_diagnostic().fields.target, "Missing", "diagnostic fields detached")
local detached_diagnostic_json = diagnostic:to_json()
detached_diagnostic_json.fields.target = "Mutated again"
check_equal(failed:compilation_diagnostic().fields.target, "Missing", "diagnostic JSON detached")

local parse_failure = linkedspec.semantic_index(
  "Top:::\n /x/\n",
  options("parse-failure.spec", "identity")
)
check_same_json(parse_failure:compilation_authority():to_json(), json.harray({
  parsed = false,
  validated = false,
  compiled = false,
}), "parse failure authority")
check_equal(parse_failure:compilation_diagnostic().code, "semantic_index_parse_failed",
  "parse fallback code")
check_equal(parse_failure:compilation_diagnostic().stage, "parse_source", "parse fallback stage")
check_equal(parse_failure:compilation_diagnostic().fields.line, 1, "parse fallback line")

local empty = linkedspec.semantic_index("", options("empty.spec", "identity"))
check_same_json(empty:compilation_diagnostic():to_json(), json.harray({
  code = "no_rules_defined",
  stage = "validate_spec",
  message = "spec does not define any rules",
  fields = json.harray(),
}), "empty native diagnostic")

local missing_entry = linkedspec.semantic_index(
  "Top::\n /x/\n",
  options("missing-entry.spec", "identity", "Missing")
)
check_same_json(missing_entry:compilation_authority():to_json(), json.harray({
  parsed = true,
  validated = true,
  compiled = false,
}), "missing entry authority")
check_same_json(missing_entry:compilation_diagnostic():to_json(), json.harray({
  code = "entry_rule_not_found",
  stage = "select_entry_rule",
  message = "entry rule 'Missing' is not defined",
  fields = json.harray({ entry_rule = "Missing" }),
}), "missing entry diagnostic")

local duplicate = linkedspec.semantic_index(
  "Top:\n /a/\nTop:\n /b/\n",
  options("duplicate.spec", "identity")
)
check_equal(duplicate:compilation_diagnostic().code, "semantic_index_validation_failed",
  "non-portable validation fallback code")
check_equal(duplicate:compilation_diagnostic().stage, "validate_source",
  "non-portable validation fallback stage")
check_equal(duplicate:compilation_diagnostic().message, "duplicate rule label 'Top'",
  "non-portable validation fallback message")

local compile_failure = linkedspec.semantic_index(
  "Top::\n /x/ E { return(array(items)) }\n", -- selector-rejection fixture
  options("compile-failure.spec", "identity")
)
check_equal(compile_failure:compilation_authority().validated, true, "compile fallback validated")
check_equal(compile_failure:compilation_authority().compiled, false, "compile fallback compiled absence")
check_equal(compile_failure:compilation_diagnostic().code, "semantic_index_compilation_failed",
  "compile fallback code")
check_equal(compile_failure:compilation_diagnostic().stage, "compile_source", "compile fallback stage")
check_contains(compile_failure:compilation_diagnostic().message, "aggregate_selector_removed",
  "compile fallback message")

local original_plan_builder = source_emitter.build_generated_rule_plan
source_emitter.build_generated_rule_plan = function()
  error(source_emitter.generated_source_error({
    stage = source_emitter.VALIDATE_GENERATED_PLAN_STAGE,
    code = source_emitter.GENERATED_PLAN_ROW_COUNT_MISMATCH_CODE,
    summary = "synthetic generated-plan failure",
    source_identity = "synthetic.spec",
  }), 0)
end
local plan_failure = linkedspec.semantic_index(graph, options("plan-failure.spec", "identity"))
source_emitter.build_generated_rule_plan = original_plan_builder
check_equal(plan_failure:compilation_diagnostic().code, "semantic_index_generated_plan_failed",
  "plan fallback code")
check_equal(plan_failure:compilation_diagnostic().stage, "build_generated_plan", "plan fallback stage")
check_equal(plan_failure:compilation_diagnostic().message, "synthetic generated-plan failure",
  "plan fallback message")

local sentinel = {}
source_emitter.build_generated_rule_plan = function() error(sentinel, 0) end
local sentinel_ok, sentinel_error = pcall(function()
  return linkedspec.semantic_index(graph, options("sentinel.spec", "identity"))
end)
source_emitter.build_generated_rule_plan = original_plan_builder
check_equal(sentinel_ok, false, "unrecognized plan sentinel rejected")
check_equal(sentinel_error, sentinel, "unrecognized plan sentinel identity")

local target = linkedspec.semantic_index(
  "Top::\n { fail(\"target must not run\") }\n",
  options("target.spec", "identity")
)
check_equal(target:semantic_snapshot().state, "compiled", "target source compiles without execution")
check_equal(target:semantic_snapshot().has_execution, false, "target execution remains absent")
check_equal(target:generated_plan_input().rows[1].family, "default", "target plan family")

local opaque_values = {
  { snapshot, "SemanticSnapshot", "snapshot" },
  { authority, "SemanticCompilationAuthority", "authority" },
  { diagnostic, "SemanticCompilationDiagnostic", "diagnostic" },
  { entry, "SemanticEntrySelection", "entry" },
  { plan, "SemanticGeneratedPlanInput", "plan" },
}
for _, case in ipairs(opaque_values) do check_opaque(case[1], case[2], case[3]) end

local public_json = {
  snapshot:to_json(),
  authority:to_json(),
  diagnostic:to_json(),
  entry:to_json(),
  plan:to_json(),
}
for index, value in ipairs(public_json) do check(plain_json(value), "public JSON plain " .. index) end
local forbidden_keys = {
  ast = true,
  body_ast = true,
  compiled_spec = true,
  definition_order = true,
  descriptor = true,
  path = true,
  rules_by_label = true,
  source = true,
  source_bytes = true,
  source_text = true,
}
local found_keys = collect_keys(public_json)
for key in pairs(forbidden_keys) do check_equal(found_keys[key], nil, "public JSON forbids " .. key) end
check_equal(graph_index.semantic_query, nil, "semantic query remains absent")
check_equal(graph_index.semantic_records, nil, "semantic records remain absent")

local implementation = read_file("lua/src/linkedspec/semantic_compilation_outcome.lua")
for token, expected in pairs({
  ["user_function_definition_parser.parse_spec_with_staged_user_function_definitions"] = 1,
  ["spec_validator.validate_spec"] = 1,
  ["compiled_spec.compile_spec"] = 1,
  ["compiled_spec.resolve_entry_rule"] = 1,
  ["source_emitter.build_generated_rule_plan"] = 1,
}) do
  local count = 0
  local position = 1
  while true do
    local found = implementation:find(token, position, true)
    if found == nil then break end
    count = count + 1
    position = found + #token
  end
  check_equal(count, expected, "single outcome call " .. token)
end
for _, forbidden in ipairs({
  "linkedspec.spec_loader",
  "runtime_parse(",
  "runtime_execute(",
  "execute_generated_parser",
  "diagnostic_sink",
  "semantic_observation",
  "to_descriptor_json",
}) do
  check_equal(implementation:find(forbidden, 1, true), nil, "outcome source forbids " .. forbidden)
end
check(implementation:find("authored_definitions = definitions", 1, true) ~= nil,
  "merged authored definitions retained")

if #failures > 0 then error(table.concat(failures, "\n"), 0) end
io.stdout:write("Lua semantic-index compilation foundation: ", assertions, " assertions passed\n")
